version-published/checkbox_version_published.py

```diff
@@ -0,0 +1,346 @@
+        url = (
+            f"http://ppa.launchpad.net/checkbox-dev/{package_spec.channel}"
+            f"/ubuntu/pool/main/c/{package_spec.source}"
+            f"/{package_spec.package}_{package_spec.version}~ubuntu"
+            f"{package_spec.ubuntu_version}.1_{package_spec.arch}.deb"
+        )
+        packages_available[package_spec] = url_header_check(url)
+
+    # Print the list of packages that were not found.
+    not_available = [
+        package_spec
+        for package_spec, is_available in packages_available.items()
+        if not is_available
+    ]
+    if not_available:
+        print("Not all packages were available in the PPA.")
+        print("Here is the list of packages that were not found:")
+        print("name                                   | ubuntu version | arch")
+        print("-" * 66)
+        for package_spec in not_available:
+            print(
+                f"{package_spec.package:<38} | "
+                f"{package_spec.ubuntu_version:<14} | "
+                f"{package_spec.arch:<5}"
+            )
+        print("-" * 66 + "\n")
+    else:
+        print("All packages were found.")
+
+
+def check_availability(
+    snap_specs: list, package_specs: list, timeout: int
+) -> None:
+    # Dict to store whether each snap and package is available.
+    snaps_available = {snap_spec: False for snap_spec in snap_specs}
+    packages_available = {
+        package_spec: False for package_spec in package_specs
+    }
+    # Set the deadline.
+    deadline = time.time() + timeout
+    while True:
+        if not all(snaps_available.values()):
+            check_snaps_availability(snap_specs, snaps_available)
+        if not all(packages_available.values()):
+            check_packages_availability(package_specs, packages_available)
+
+        # Exit the loop if all snaps and/or packages are found.
+        if all(snaps_available.values()) and all(packages_available.values()):
+            break
+
+        # Exit the loop if the timeout is reached.
+        if time.time() > deadline:
+            raise TimeoutError("Timeout reached.")
+
+        print("--- Waiting 30 seconds before retrying ---\n")
+        # Wait before the next iteration.
+        time.sleep(30)
+    print("All snaps/packages for the specific version were found.")
+
+
+def main(argv):
+    parser = argparse.ArgumentParser(
+        description="Check whether snaps are available in the snap store."
+    )
+    parser.add_argument("version", help="Version of checkbox to check for.")
+    parser.add_argument(
+        "checkbox_yaml",
+        type=argparse.FileType("r"),
+        help="Path to the YAML file specifying the snap requirements.",
+    )
+    parser.add_argument(
+        "--timeout",
+        help="Timeout in seconds after which the program will stop checking.",
+        default=300,
+        type=float,
+    )
+    args = parser.parse_args(argv[1:])
+
+    yaml_content = yaml.load(args.checkbox_yaml, Loader=yaml.FullLoader)
```

Comment:
If the yaml is empty, the loader will return None and the code will crash, it may be worth checking if it's a dict.